sw/device/lib/dif/dif_hmac.c

```diff
@@ -0,0 +1,329 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/dif/dif_hmac.h"
+
+#include "hmac_regs.h"  // Generated.
+#include "sw/device/lib/base/memory.h"
+#include "sw/device/lib/base/mmio.h"
+
+#define HMAC_HASH_SIZE_BITS 256u
+
+/**
+ * Load |key| into |hmac|.
+ */
+static void load_key(const dif_hmac_t *hmac, const uint8_t *key) {
+  for (int i = 0; i < HMAC_PARAM_NUMWORDS; ++i) {
+    ptrdiff_t offset = i * sizeof(uint32_t);
+    uint32_t key_part;
+
+    memcpy(&key_part, key + offset, sizeof(key_part));
+
+    mmio_region_write32(hmac->base_addr, HMAC_KEY0_REG_OFFSET + offset,
+                        key_part);
+  }
+}
+
+/**
+ * Read the status register from |hmac|.
+ *
+ * @param hmac The HMAC device to read the status register from.
+ * @return The contents of hmac.STATUS.
+ */
+static uint32_t get_status(const dif_hmac_t *hmac) {
+  return mmio_region_read32(hmac->base_addr, HMAC_STATUS_REG_OFFSET);
+}
+
+/**
+ * Returns the number of entries in the FIFO of |hmac|. If the FIFO is empty,
+ * this function will return 0, and if the FIFO is full, this funciton will
+ * return |HMAC_FIFO_MAX|.
+ *
+ * @param hmac The HMAC device to check the FIFO size of.
+ * @return The number of entries in the HMAC FIFO.
+ */
+static uint32_t get_fifo_entry_count(const dif_hmac_t *hmac) {
+  return (get_status(hmac) >> HMAC_STATUS_FIFO_DEPTH_OFFSET) &
+         HMAC_STATUS_FIFO_DEPTH_MASK;
+}
+
+/**
+ * A helper funciton for calculating |HMAC_FIFO_MAX| - |get_fifo_entry_count()|.
+ */
+static uint32_t get_fifo_available_space(const dif_hmac_t *hmac) {
+  return HMAC_MSG_FIFO_SIZE_WORDS - get_fifo_entry_count(hmac);
+}
+
+dif_hmac_result_t dif_hmac_init(const dif_hmac_config_t *config,
+                                dif_hmac_t *hmac) {
+  // Basic sanity checks on parameters. In |kDifHmacModeHmac| mode a key is
+  // required.
+  if (config == NULL || hmac == NULL ||
+      (config->key == NULL && config->mode == kDifHmacModeHmac)) {
+    return kDifHmacBadArg;
+  }
+
+  hmac->base_addr = config->base_addr;
+
+  // Clear the config, stopping the SHA engine.
+  mmio_region_write32(config->base_addr, HMAC_CFG_REG_OFFSET, 0);
+
+  // Clear INTER_STATE.
+  mmio_region_write32(config->base_addr, HMAC_INTR_STATE_REG_OFFSET,
+                      (1 << HMAC_INTR_ENABLE_HMAC_DONE) |
+                          (1 << HMAC_INTR_STATE_FIFO_EMPTY) |
+                          (1 << HMAC_INTR_STATE_HMAC_ERR));
+
+  uint32_t device_config = 0;
+
+  // Set the byte-order of the input message.
+  switch (config->message_endianness) {
+    case kDifHmacEndiannessBig:
+      device_config |= (1 << HMAC_CFG_ENDIAN_SWAP);
```

Comment:
This depends on host endianness, so it's not 100% accurate to say that we have to swap endianness to be big-endian. That being said, without doing gross pointer stuff or depending on #defines to give us this information I don't know that we can be fully accurate here. Can you add a TODO in here to address that? We can commit with the TODO I just want to make sure we fix it up later.